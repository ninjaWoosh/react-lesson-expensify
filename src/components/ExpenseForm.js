import React, {Component} from "react";
import moment from "moment";
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends Component{

    constructor(props){
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : "",
            amount: props.expense ? (props.expense.amount/100).toString() : "",
            note: props.expense ? props.expense.note : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ""
        };
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({amount}));
        }       
    };

    onDateChange = (createdAt) => {
        if(createdAt)
        this.setState(() => ({createdAt}));
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            this.setState(() => ({error: "Please provide description and amount"}));
        }else{
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };

    render(){
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        value={this.state.description}
                        placeholder="Description" 
                        autoFocus
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        onChange={this.onAmountChange}
                        value={this.state.amount}
                        type="text" 
                        placeholder="Amount" 
                    />
                    <SingleDatePicker
                        onDateChange={this.onDateChange}
                        date={this.state.createdAt}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                    />
                    <textarea 
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        placeholder="Add notes here (Optional)">
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}