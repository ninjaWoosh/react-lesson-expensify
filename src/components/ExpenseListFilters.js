import React from "react";
import 'react-dates/initialize';
import {DateRangePicker} from "react-dates";
import 'react-dates/lib/css/_datepicker.css';
import {connect} from "react-redux";
import {setFilterText, sortByAmount, sortByDate, setStartDate, setEndDate} from "../actions/filters";


export class ExpenseListFilters extends React.Component{
    state = {
        calendarFocused:null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    };

    onTextChange = (e) => {
        this.props.setFilterText(e.target.value);
    };

    onSortChange = (e) => {
        if(e.target.value === "date"){
            this.props.sortByDate();
        }else if(e.target.value === "amount"){
            this.props.sortByAmount();
        }
    };

    render(){
        return (
                <div>
                    <input type="text" 
                        value={this.props.filters.text}
                        onChange={this.onTextChange}/>

                    <select 
                        value={this.props.filters.sortBy}
                        onChange={this.onSortChange}>
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                    </select>
                    <DateRangePicker
                        startDateId="startDate"
                        startDate={this.props.filters.startDate}
                        endDateId="endDate"
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                </div>
            );
    }
}

const mapStateToProps = (state) => (
    {
        filters: state.filters
    }
); 

const mapDispatchToProps = (dispatch) => ({
    setFilterText: (text) => dispatch(setFilterText(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);