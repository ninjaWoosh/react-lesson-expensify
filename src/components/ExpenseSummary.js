import React from 'react';
import selectExpenseTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import { connect } from "react-redux";
import numeral from "numeral";

export const ExpenseSummary = ({count, total}) => (
    <div>
        <p>Viewing {count} expenses totalling {total}</p>
    </div>
);


const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters);
    const count = expenses.length;
    const total = numeral(selectExpenseTotal(expenses) / 100).format("$0,0.00");

    return {
        count,
        total
    };
};

export default connect(mapStateToProps)(ExpenseSummary);