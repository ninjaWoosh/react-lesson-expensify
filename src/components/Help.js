import React from "react";

const Help = () => (
    <p> This is the Help!</p>
);

export default Help;