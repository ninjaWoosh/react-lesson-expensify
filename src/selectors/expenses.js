import moment from "moment";

export default (expenses, {text, sortBy, startDate, endDate}) => {
    
    return expenses.filter((expense) => { 
        const createdAtMoment = moment(expense.createdAt);
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = startDate ? moment(startDate).isSameOrBefore(createdAtMoment): true;
        const endDateMatch = endDate ? moment(endDate).isSameOrAfter(createdAtMoment) : true;

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }

        if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1;
        }
    });
}