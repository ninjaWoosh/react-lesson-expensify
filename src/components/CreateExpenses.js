import React from "react";
import ExpenseForm from "./ExpenseForm";
import {connect} from "react-redux";
import {addExpense} from "../actions/expenses";

export class CreateExpenses extends React.Component{
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push("/");
    };

    render(){
        return (
            <div>
            <h1>Create new expense</h1>
            <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    };
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(CreateExpenses);