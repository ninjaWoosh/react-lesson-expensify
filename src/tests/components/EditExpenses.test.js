import React from 'react';
import {shallow} from "enzyme";
import {EditExpenses} from "../../components/EditExpenses";
import expenses from "../fixtures/expenses";

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpenses expense={expenses[0]} editExpense={editExpense} removeExpense={removeExpense} history={history} />);
});

test("Should render edit expenses component", () => {
    expect(wrapper).toMatchSnapshot();
});

test("Should render with correct submit data", () => {
    wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
    expect(editExpense).toBeCalledWith(expenses[0].id, expenses[0]);
    expect(history.push).toBeCalledWith("/");
});

test("Should render with click data", () => {
    wrapper.find("button").prop("onClick")();
    expect(removeExpense).toBeCalledWith(expenses[0].id);
    expect(history.push).toBeCalledWith("/");
});