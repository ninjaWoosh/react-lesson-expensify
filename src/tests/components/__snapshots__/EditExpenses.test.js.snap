// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expenses component 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 123,
        "createdAt": 0,
        "description": "Coffee",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
